### PR TITLE
added replicas count for daemonsets to prevent massive pod eviction

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
#### Which component this PR applies to?

vertical-pod-autoscaler
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
This PR adds ability to count "replicas" in daemonsets, to prevent massive one-time pod restart  in daemonsets.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Added support for counting replicas in deamonsets to limit how fast VPA evicts them.
```
NONE
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by nabokihms** [comment]
Hello, @krzysied @jbartosik. What can we do to move this PR further? It seems stuck without comments, and it is unceratin what to do next. We will appreciate any instructions.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go [hunk 1]
```diff
@@ -25,13 +25,14 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
 	appsinformer "k8s.io/client-go/informers/apps/v1"
 	coreinformer "k8s.io/client-go/informers/core/v1"
 	kube_client "k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/record"
 	klog "k8s.io/klog/v2"
+
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
```
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go:35]: Please don't move this line

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go [hunk 2]
```diff
@@ -26,12 +26,13 @@ import (
...
 	apiv1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
-	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
 	appsinformer "k8s.io/client-go/informers/apps/v1"
 	coreinformer "k8s.io/client-go/informers/core/v1"
 	"k8s.io/client-go/kubernetes/fake"
 	"k8s.io/client-go/tools/cache"
+
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
```
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go:34]: Please don't move this line.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go [hunk 3]
```diff
@@ -26,12 +26,13 @@ import (
...
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/test"
```
Comment (jbartosik) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go:35]: Please this import in group with other "k8s.io/" imports

**Review by jbartosik — CHANGES_REQUESTED** [review]
I left some small comments for code.

From description of the PR it sounds like there will be an user-visible change to behavior of VPA with regards to DaemonSets.

Can you give an example of situation where this improves VPAs behavior? (it might be worth adding a e2e test for this).

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go (hunk 2 above)
Comment (RomanenkoDenys) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go]: Sorry, it is Goland. Fixed.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go (hunk 3 above)
Comment (RomanenkoDenys) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction_test.go]: Sorry, it is Goland. Fixed.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go (hunk 1 above)
Comment (RomanenkoDenys) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Sorry, it is Goland. Fixed.

Comment (RomanenkoDenys) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Sorry, it is Goland. Fixed.

**Comment by jbartosik** [comment]
Hi, I see that you addressed my comments for small code clean ups.

Could you:

1) Squash commits in this PR and
2) Add a release note; In "Does this PR introduce a user-facing change?" put something like "added support for counting replicas in deamonsets to limit how fast VPA evicts them"?

**Comment by RomanenkoDenys** [comment]
> Hi, I see that you addressed my comments for small code clean ups.
> 
> Could you:
> 
> 1. Squash commits in this PR and
> 2. Add a release note; In "Does this PR introduce a user-facing change?" put something like "added support for counting replicas in deamonsets to limit how fast VPA evicts them"?

Hi, @jbartosik. I did what you asked, thanks for the reply!

**Review by jbartosik — APPROVED** [review]
/lgtm
/approve

**Pull request merged**